commit 0e86ee934b6587dc288e756eb980da002bcbd240
Author: Alex Richardson <arichardson.kde@gmail.com>
Date:   Mon Sep 3 10:16:48 2018 +0100

    MacOS: Add configure option for additional QStandardPaths data directories
    
    On MacOS QStandardPaths::GenericDataLocation only contains
    "~/Library/Application Support" and "/Library/Application Support".
    Package managers such as homebrew or KDE craft install into a custom
    prefix and do not/cannot write to either of these directories. This makes
    it impossible for them to share the generic data between different
    applications. The current workaround used e.g. by homebrew-kde is to create
    symlinks for every file/directory that is needed in /usr/local/share/ in
    the users "~/Library/Application Support" directory. However, this causes
    issues if you forget to create the appropriate link and makes it difficult
    for the user to override a data file from the system-wide installation with
    a file in the home directory.
    
    This commit introduces a new hidden ./configure option
    `-macos-additional-datadirs`. This option is a colon-separated list of
    additional directories that QStandardPaths searches before looking in
    "/Library/Application Support".
    
    I also introduced a new environment variable QT_DATA_DIRS that can be
    used to override this ./configure option.
    TODO: should this be QT_MACOS_DATA_DIRS instead?
    
    Change-Id: I23b046fcc9520181286fe300827645cd608b0201

diff --git a/src/corelib/io/qstandardpaths_mac.mm b/src/corelib/io/qstandardpaths_mac.mm
index 11b5cc8c37..da91ec3ae5 100644
--- a/src/corelib/io/qstandardpaths_mac.mm
+++ b/src/corelib/io/qstandardpaths_mac.mm
@@ -184,6 +184,31 @@ QStringList QStandardPaths::standardLocations(StandardLocation type)
         dirs << writableLocation(PicturesLocation) << QLatin1String("assets-library://");
 #endif
 
+#if !defined(QT_BOOTSTRAPPED) && defined(Q_OS_MACOS)
+    if (type == AppDataLocation || type == AppLocalDataLocation || type == GenericDataLocation) {
+        QStringList extraDataDirs;
+        // Add the value of -macos-generic-datadirs configure argument to the list.
+        // This allows e.g. homebrew to use a Qt build that is configured to search
+        // for generic data in /usr/local/share instead of requiring it to be in
+        // /Library/Application Support which is not writable from the homebrew sandbox.
+#ifdef QT_MACOS_ADDITIONAL_DATADIRS
+        QString fromConfigure = QStringLiteral(QT_MACOS_ADDITIONAL_DATADIRS);
+        extraDataDirs = fromConfigure.split(QLatin1Char(':'), QString::SkipEmptyParts);
+#endif
+        // Also allow the environment variable QT_DATA_DIRS to override this default
+        QByteArray qtDataDirsEnv = qgetenv("QT_DATA_DIRS");
+        if (!qtDataDirsEnv.isEmpty()) {
+            extraDataDirs = QFile::decodeName(qtDataDirsEnv).split(QLatin1Char(':'), QString::SkipEmptyParts);
+        }
+        if (type == AppDataLocation || type == AppLocalDataLocation) {
+            for (QString &dir : extraDataDirs) {
+                appendOrganizationAndApp(dir);
+            }
+        }
+        dirs << extraDataDirs;
+    }
+#endif
+
     if (type == GenericDataLocation || type == FontsLocation || type == ApplicationsLocation
             || type == AppDataLocation || type == AppLocalDataLocation
             || type == GenericCacheLocation || type == CacheLocation) {
